[proposal] RequestData|ResponseData API migration (#116)
kroxylicious/design: proposals/116-kafka-api-migration.md

Kafka does not consider these packages to be stable public APIs. The project has no obligation to maintain backward compatibility, and further reorganisations are likely. Because these unstable Kafka classes are exposed through our public API, we cannot simply absorb these changes beneath our abstraction layer. Instead, every upstream reorganization forces a breaking change directly onto **Filter Authors** who rely on the Filter API to implement their own custom business logic.

For a 1.0 release, Kroxylicious needs a stable public API surface that the project controls. The Kafka protocol wire format itself is stable (governed by KIPs), but the Java classes representing that format are not.

## Proposal

### Scope: own the full API surface in one move

Every type that appears in Kroxylicious's public API should live in a Kroxylicious namespace. This means a single breaking change for filter developers - one import migration - rather than a drip of repeated breakage as Kafka reorganises its internals.

> "Own the surface" does not mean "rewrite everything from scratch." The copied source files retain their original Apache Software Foundation copyright headers - the ASF remains the author of the underlying code, and these files are redistributed under the same Apache 2.0 licence. The underlying implementations can be replaced with Kroxylicious-native ones incrementally, gated by the verification harness proving each replacement is correct.

Reviewer: "The underlying implementations can be replaced with Kroxylicious-native ones incrementally" they _can_ be, but we should be clear that that's done something that's necessary or part of this proposal.

Author: Done in [`7df5748` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/7df5748122205c32bece1d9b0ce6aaa2cc6622ae)

### Approach: own the source from the start

> Kafka's `MessageGenerator` (~35 files in `generator/src/main/java/org/apache/kafka/message/`) is copied into the Kroxylicious repo as a new module (`kroxylicious-kafka-message-generator`). It is invoked during the build against the upstream JSON IDL spec files (bundled inside `kafka-clients`) to produce `*Data` source files under the Kroxylicious namespace (`io.kroxylicious.kafka.*`). The non-generated classes - protocol infrastructure, record classes, and scattered `common.*` types - are similarly copied into the Kroxylicious repo as source under the same namespace, retaining their original Apache Software Foundation copyright headers. All of these classes live in a new `kroxylicious-kafka-common` module, mirroring the structure of Kafka's own `org.apache.kafka.common` package.

Reviewer 2: nit - naming: you say namespace in this sentence when you mean java package.
also this paragraph would be clearer if:
a) make the forking of MessageGenerator into a module kroxylicious-kafka-message-generator its own paragraph.
b) mention the role of the new module `kroxylicious-kafka-common` and make it clear that it will invoke the MessageGenerator

Author: Done in [`207c48b` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/207c48b1ba7a3e8b146949154c6a47abe05d7ae8)

This gives full source ownership from day one. Every upstream change is a deliberate choice - review it, absorb it, or skip it. Generator enhancements (type-safe request-response pairing, Javadoc from IDL, `aliases`) are first-class commits to the in-repo generator, not patches layered on top. And when the time comes to replace buffer/record classes with Netty-native implementations, there is no bytecode transformation to unwind - we already own the source.

> The `kroxylicious-krpc-plugin` currently depends on `kafka-clients` solely for `ApiKeys`. After this change it depends on the new artifact instead (or we could codegen an `ApiKeys` from the RPC definitions themselves). Its role in generating filter interfaces, invokers, and decoders is otherwise unchanged.

Reviewer 2: please disambiguate "new artefact".  just naming the module is simplest.

Author: Done in [`207c48b` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/207c48b1ba7a3e8b146949154c6a47abe05d7ae8)

> The `kroxylicious-krpc-plugin` currently depends on `kafka-clients` solely for `ApiKeys`. After this change it depends on the new artifact instead (or we could codegen an `ApiKeys` from the RPC definitions themselves). Its role in generating filter interfaces, invokers, and decoders is otherwise unchanged.

Reviewer 2: > After this change [kroxylicious-krpc-plugin] depends on the new artifact instead (or we could codegen an `ApiKeys` from the RPC definitions themselves). Its role in generating filter interfaces, invokers, and decoders is otherwise unchanged.

Either way it sounds a bit ugly.  

 I *think* eliminating the use of `ApiKeys` class from `KrpcGenerator` mechanism itself will simplify the problem.  From a quick look, I think its only use-case is mapping id to name.  The Kafka RPC definitions include the ids and names, so I think it should be an easy refactor.

```json
{
  "apiKey": 1,
  "type": "request",
  "listeners": ["broker", "controller"],
  "name": "FetchRequest"
}
```

Reviewer 3: Good idea, but I'd keep it out of scope for this proposal unless it simplifies the package move itself. If eliminating the `ApiKeys` dependency from KrpcGenerator makes the migration easier — fewer things to relocate, fewer cross-module dependencies to untangle — then it's worth doing as part of this work. If it's just a nice cleanup, it can come after.

Reviewer 2: > If eliminating the ApiKeys dependency from KrpcGenerator makes the migration easier — fewer things to relocate, fewer cross-module dependencies to untangle — then it's worth doing as part of this work. If it's just a nice cleanup, it can come after.

My intuition is that it is the former.    It should be a change we can deliver separately, which then simplifies the rest of the problem.  Let me see if I can brush up a simple PR to test if my intuition is correct.

Reviewer 2: I've put up a draft PR exploring this idea: https://github.com/kroxylicious/kroxylicious/pull/4472

Reviewer 4: I think this is worth mentioning in the proposal since it breaks a potential cycle. kroxylicious-api depends on the generator to build the Filter interfaces, so we want to avoid krpc generator depending on kroxylicious-api. (think we've got a consensus that folding all the new bits into kroxylicious-api is the preference now)

Reviewer 2: I've marked https://github.com/kroxylicious/kroxylicious/pull/4472 as ready for review.  If it is accepted, I suggest we merge that and rebase this proposal.

Author: Removed some mentions of `ApiKeys` [`ed16ac9` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/ed16ac996d0bea794c762f8e0f22344842dd3b58) + [`d125e3d` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/d125e3d9037be605be8b38ac0e4d2a7774176623)

### What gets copied and generated

| Component                                  | Source                                    | Approach                                               |
|--------------------------------------------|-------------------------------------------|--------------------------------------------------------|

> | `MessageGenerator` + friends (~35 files)   | `kafka/generator/src/`                    | Copied into Kroxylicious repo as a new module          |

Reviewer: "as a new module": let's decide what it's called.

Author: Decided on `kroxylicious-kafka-message-generator` for the generator and `kroxylicious-kafka-common` for the rest (we copy Kafka's package structure), open to suggestions though :smiley:.

[`40a6db8` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/40a6db8176d6ac1f16edaca3559817c93467636c)

> | `MessageGenerator` + friends (~35 files)    | `kafka/generator/src/`                    | Copied into `kroxylicious-kafka-message-generator`     |

Reviewer 2: The Kafka classes won't adhere to our coding conventions, so I'd bet money checkstyle and errorprone are going to light up.   I suggest we relax the rules for the copied file locations and have a separate task to make them adhere to our own conventions.

Author: Mentioned in [`207c48b` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/207c48b1ba7a3e8b146949154c6a47abe05d7ae8)

> | `*Data` classes (~198)                      | IDL specs + copied `MessageGenerator`     | Generated into `kroxylicious-kafka-common`             |

Reviewer 2: So `kroxylicious-api` is going to depend on `kroxylicious-kafka-common`.   Curious - did you consider running the generator in kroxylicious-api instead?  I'm wondering if we really need the kroxylicious-kafka-common module?

Author: I decided against it because `kroxylicious-api` looks to entirely consist of interfaces and very simple/short classes, so I thought putting implementation-heavy code in there would be a bad idea. Let me know if this is an incorrect assumption though.

Reviewer 2: We already have implementation heavy code in this module (the generated *Filter classes) generated by the krpc-plugin.  What's the difference?   I think keeping the number of public facing modules small is advantageous -  modules are part of the public API too.

We've also just learnt today that more module leads to MC limits problems too https://github.com/kroxylicious/kroxylicious/issues/4467

Reviewer 4: I agree with starting with it in `kroxylicious-api`, this contains the API filters need to implement which will include all the forked classes down to MemoryRecords and so on.

One argument for keeping them separate is that other technology could take advantage of this library if they want a Kafka serialization library with public-API type compatibility guarantees. Very niche so I think we'd be better starting with them in `kroxylicious-api` and could extract later into independent module if anyone ever asked. (or they can just depend on api and live with a few irrelevant classes, it's not a big footprint)

Reviewer 3: Good challenge. I want to own something first: I previously proposed `kroxylicious-kafka-protocol` as a separate module from `kroxylicious-api` and argued for it on cohesion grounds — "everything in this module exists to represent or operate on the Kafka wire protocol." I still think that's a coherent principle, but cohesion alone isn't enough justification for a module boundary. The question is whether anything has a reason to depend on the protocol types *without* also depending on `kroxylicious-api`, and I can't find a convincing case for that.

The MC limits argument cuts both ways for me — I agree they're a real problem, but I don't think "fewer modules" is the right fix on its own. Modules should have a single reason to change. The issue is that the protocol types and the filter API types *don't* have separate reasons to change — they move together. So the boundary isn't load-bearing.

I think you're right: fold them together.

Author: Folded `kroxylicious-kafka-common` into `kroxylicious-api`: [`8ae28fc` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/8ae28fc98f35881f6e60a6b3f84d1452953774c1).

(Kept `kroxylicious-kafka-message-generator` and `kroxylicious-kafka-common-tests`)

Exploratory work on the [`piotrpdev/kroxylicious#refactor/generate-data-class`](https://github.com/piotrpdev/kroxylicious/tree/refactor/generate-data-class) branch built a prototype of this infrastructure covering all 198 specs: parameterised fidelity tests at every valid version, randomised field-value tests (including explicit coverage of `null` for nullable fields), jqwik property-based tests for full-depth message instances, and a compile-time spec count guard. That prototype was designed around the FreeMarker template approach (see ["Rejected Alternatives"](#rejected-alternatives)), where generated classes were compiled at test time via `ToolProvider.getSystemJavaCompiler()` and loaded via `URLClassLoader`. With the source-ownership approach the classes are already on the classpath at build time, so only the class-loading mechanism changes - the verification logic itself carries over unchanged.

> These tests live in the Kroxylicious repo alongside the generated and copied classes, with `kafka-clients` as a test-scoped dependency. They run as part of the normal CI pipeline - when `kafka-clients` is bumped, they immediately report any incompatibility.

Reviewer 2: What module do this tests live in?

Author: Mentioned in [`207c48b` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/207c48b1ba7a3e8b146949154c6a47abe05d7ae8)

| New `validVersions` max           | Low        | Just more versions to test                  |
| New field with `ignorable: true`  | Low        | Safe at all versions                        |

#### Proactive early warning

Dependabot opens a PR bumping `kafka-clients` to a new version on release automatically. CI runs on the PR - regenerating `*Data` classes and running the fidelity tests and `japicmp` comparison against the new `kafka-clients` (see section below for details on tests).

If CI passes, the PR can be reviewed and merged with minimal effort. If CI fails, the PR surfaces exactly what broke - a fidelity test failure identifies a wire incompatibility; an `EXPECTED_SPEC_COUNT` failure flags a spec change requiring review. Either way, the problem is visible and actionable immediately, before anyone has to manually discover the new release.

#### Drift in source-copied classes

The non-generated classes (`MemoryRecords`, `RecordBatch`, protocol infrastructure, etc.) are copied source that may diverge from upstream over time as Kafka evolves. Drift is detected through two complementary mechanisms:

> 1. **`japicmp` API comparison:** `japicmp-maven-plugin` compares by fully-qualified class name, so we cannot point it directly at our jar and `kafka-clients` - the package names do not match. Instead, we shade `kafka-clients` with the same `org.apache.kafka` → `io.kroxylicious.kafka` relocation into a throwaway reference artifact, then run `japicmp` comparing our source-built jar against that reference. Both jars now have matching package names, and `japicmp` reports exactly which APIs have diverged from upstream - method renames, signature changes, removed types. Kroxylicious already uses `japicmp` in the main build, so the plugin infrastructure is in place.

Reviewer 2: I question whether we really need this.  The important thing is wire level compatibility.

Reviewer 3: I think wire compatibility is necessary but not sufficient on its own. These are Kafka's internal classes — Kafka doesn't consider them public API and won't reliably document changes to them. The 4.3.0 `record.*` move that motivated this proposal is a good example: even reading the JIRA (https://issues.apache.org/jira/browse/KAFKA-20128) you wouldn't pick it as a class relocation — you'd have to read the actual PR (https://github.com/apache/kafka/pull/21412) to connect the dots. No reasonable level of release note monitoring would have caught it proactively.

`japicmp` is the only automated signal for structural drift that isn't wire-visible — a renamed method, a changed signature, a removed type. Without it, the alternative is manually diffing our copies against upstream, which won't happen consistently.

That said, it's a transitional safety net. As each copied class gets replaced with a Kroxylicious-native implementation, we stop running `japicmp` against it — we've deliberately diverged at that point. But for classes that are still essentially Kafka's code, the drift signal remains valuable. The scope shrinks progressively, class by class.

Reviewer 2: > I think wire compatibility is necessary but not sufficient on its own. 

I don't see this.  Once we've made the fork the Data classes become our public API.  If Kafka refactor their Data classes to add/remove/change methods, we don't care.  The only thing we care about is that the serialised byte array they produce is mutually intelligible.

I think we risk doing a load of  japicmp work that we'll immediately consign to bin.

Reviewer 3: I get the argument — once we've forked, Kafka's API changes are irrelevant to us. My concern is narrower: if Kafka fixes a bug in one of the copied classes, how do we find out and pull it in?

I've convinced myself there's no good automated answer to that. japicmp only tracks signature changes — invisible to a logic fix inside a method body. A scheduled diff against upstream trunk breaks down the moment we deliberately don't absorb a change, because every subsequent diff is permanently dirty. Wire fidelity tests catch serialisation bugs but not behavioural ones.

So I think you're right to drop japicmp for the copied source classes — it wasn't giving us what I thought it was. The gap (non-wire logic bugs in copied source) is real but unautomatable. We should acknowledge it explicitly in the proposal and accept it as a known limitation of the source-ownership approach. The japicmp usage for the generated *Data classes is a separate matter and unaffected by this.

Reviewer 4: Another interesting case is record batch versions and compression types that are allowed per RPC version (see ProduceRequest#validateRecords), that logic won't even be in the classes we are taking ownership of. Is that validation part of the wire protocol?  By the wire protocol you can say that new compression types could pop up in any version. This is true of the current proxy code anyway, we decode at a lower level and don't use those request classes. I guess we will just need tests we can handle all compression types and keep a close eye out for new magic versions appearing in RecordBatch.

Reviewer 2: > Another interesting case is record batch versions and compression types that are allowed per RPC version (see ProduceRequest#validateRecords), that logic won't even be in the classes we are taking ownership of. Is that validation part of the wire protocol?

Oh - good point.  I think we actually have that problem today.  There's nothing to stop a ProduceResponse Filter Author generating a 'bad' batch.  https://github.com/kroxylicious/kroxylicious/issues/4489

Author: Removed `japicmp` in [`c5e7087` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/c5e70877f30d78a40555a17977c3120e14869a05)

2. **Wire-level fidelity tests:** Any change to serialisation behaviour surfaces as a test failure. These tests serialise with our classes and deserialise with the original `kafka-clients` classes (and vice versa), so any divergence in the binary format is caught immediately.

Together, `japicmp` catches API-level divergence and the fidelity tests catch wire-format divergence. Non-wire behavioural changes - validation logic, error handling - require reviewing the `kafka-clients` changelog when bumping the version, which is part of the standard upgrade process.

### Enhancements to the generator

> #### Type-safe request-response pairing

Reviewer: Personally I think this is just about worth doing. On the one hand it's a bit cute and doesn't offer a lot of extra type safety. But it also doesn't seem particularly hard or risky.

Reviewer 2: Let's keep focus on what we have to do. This proposal is big enough already.  If we decide we want to propose it separately, after this proposal is delivered, but before the release, fine.

Reviewer 3: Agree with @k-wall — this is a good idea but let's keep it out of scope for this proposal. The goal here is to take ownership of the protocol surface and unblock Kafka version upgrades. Type-safe pairing can be a follow-up once the foundation is in place.

Reviewer: > Type-safe pairing can be a follow-up once the foundation is in place.

Not without breaking compatibility again. We if we want to do this separately we'd need to land a PR for it in the same release.

Reviewer 2: >> If we decide we want to propose it separately, after this proposal is delivered, but before the release, fine.

> Not without breaking compatibility again. We if we want to do this separately we'd need to land a PR for it in the same release.

Let's aim to do this, for the same release, but let's deliver it as a follow on PR.  My motivation here is smaller steps = easier review.

Reviewer 2: Just to be clear, I'm happy for this to stay in the proposal , but lets say it will be delivered by a separate phase.

Author: Emphasized in [`0c70c80` (this PR)](https://github.com/kroxylicious/design/pull/116/commits/0c70c8090721bad0ad0bc296c01764638ca3638c)

- **~16 protocol infrastructure classes** (`org.apache.kafka.common.protocol.*`) that the generated classes depend on: `ApiMessage`, `Message`, `Readable`, `Writable`, `ObjectSerializationCache`, `MessageSizeAccumulator`, the `types.*` schema model, etc.
- **~35 record classes** (`org.apache.kafka.common.record.*`) including `MemoryRecords`, `RecordBatch`, `Record`, etc. - hand-coded implementations of Kafka's on-wire/on-disk record format.
- **`ApiKeys`** - the enum that maps API key numbers to names and version ranges.
- **Scattered `common.*` types** - `Uuid`, `ApiException` subclasses, ACL types, SASL classes, utilities.

Kroxylicious's `ByteBufAccessor` currently `implements Readable, Writable` from Kafka, bridging Netty's `ByteBuf` to Kafka's I/O abstraction. Any namespace change requires updating this bridge.

The `kroxylicious-krpc-plugin` already parses the JSON IDL specs independently using Jackson (without depending on Kafka's generator) and drives FreeMarker templates to generate filter interfaces, invokers, and decoders. Its only Kafka dependency is `ApiKeys`, used in 3 places.

## Motivation

> Kafka 4.3.0 moved 37 classes from `org.apache.kafka.common.record.*` to `org.apache.kafka.common.record.internal.*` ([KAFKA-20128](https://issues.apache.org/jira/browse/KAFKA-20128)). Upgrading Kroxylicious to 4.3.0 required import changes across 80 Java files in 9 modules. This is a symptom of a deeper problem: Kroxylicious depends on Kafka internal APIs that Kafka is free to move or remove at any time.

Reviewer 2: It is not just moves and removals that are problematic.  Changes in methods/method signatures API would be equally impactful.

Author: ```suggestion
Kafka 4.3.0 moved 37 classes from `org.apache.kafka.common.record.*` to `org.apache.kafka.common.record.internal.*` ([KAFKA-20128](https://issues.apache.org/jira/browse/KAFKA-20128)). Upgrading Kroxylicious to 4.3.0 required import changes across 80 Java files in 9 modules. This is a symptom of a deeper problem: Kroxylicious depends on Kafka internal APIs that Kafka is free to change, move or remove at any time.
```

> Kafka does not consider these packages to be stable public API. The project has no obligation to maintain backward compatibility, and further reorganisations are likely. Each one forces a breaking change on every Kroxylicious filter developer - internal and external.

Reviewer 2: By **internal** I think you you mean this project itself and the filters it ships as part of Kroxylicious.  By **external**, I think you mean the **Filter authors** i.e. users writing their own Filters against the Filter API implementing their own business logic.
 
I don't really think "internal" is really the concern.  If we didn't have a public Filter API, we could just swallow the Kafka changes and it would all be beneath the abstraction.  If we cared at all, we wouldn't need a proposal because no public API would be changd.

I think the the motivation for this change is the second group you called external.  We've defined the persona name as 'Filter Authors'.

Author: ```suggestion
Kafka does not consider these packages to be stable public APIs. The project has no obligation to maintain backward compatibility, and further reorganisations are likely. Because these unstable Kafka classes are exposed through our public API, we cannot simply absorb these changes beneath our abstraction layer. Instead, every upstream reorganization forces a breaking change directly onto **Filter Authors** who rely on the Filter API to implement their own custom business logic.
```

For a 1.0 release, Kroxylicious needs a stable public API surface that the project controls. The Kafka protocol wire format itself is stable (governed by KIPs), but the Java classes representing that format are not.

## Proposal

### Scope: own the full API surface in one move

Every type that appears in Kroxylicious's public API should live in a Kroxylicious namespace. This means a single breaking change for filter developers - one import migration - rather than a drip of repeated breakage as Kafka reorganises its internals.

> "Own the surface" does not mean "rewrite everything from scratch." The shaded and generated artifacts redistribute `kafka-clients` bytecode under the same Apache 2.0 licence. The original Apache Software Foundation copyright notices are preserved in the bytecode. The underlying implementations can be replaced with Kroxylicious-native ones incrementally, gated by the verification harness proving each replacement is correct.

Reviewer 2: nit: the licence is comment in the source, so does not end up in the bytecode.

Author: ```suggestion
"Own the surface" does not mean "rewrite everything from scratch." The shaded and generated artifacts redistribute `kafka-clients` bytecode under the same Apache 2.0 licence. The original Apache Software Foundation copyright notices are preserved. The underlying implementations can be replaced with Kroxylicious-native ones incrementally, gated by the verification harness proving each replacement is correct.
```

### Approach: shade `kafka-clients` via a dedicated repo

> The Maven Shade Plugin takes compiled classes from `kafka-clients` and republishes them under a Kroxylicious namespace. All cross-references between relocated classes are rewritten automatically at the bytecode level. This means Kafka package reorganisations - including moves to an `internal` subpackage like KAFKA-20128 - are completely invisible.

Reviewer 2: Shading can handle repackaging, but AFAIK, can't handle things like method renames or method signature changes.

Author: That's where we could use the patches to rename methods or add overloads for them

Author: ```suggestion
The Maven Shade Plugin takes compiled classes from `kafka-clients` and republishes them under a Kroxylicious namespace. All cross-references between relocated classes are rewritten automatically at the bytecode level. This means Kafka package reorganisations - including moves to an `internal` subpackage like KAFKA-20128 - are completely invisible. Breaking code changes such as method renames or signature changes are addressed by adding patches that overload the methods or wrap them appropriately.
```

> A dedicated repo in the Kroxylicious GitHub org (`kroxylicious/kafka-patches`) stores the shading configuration and any generator patches needed for future enhancements. When generator enhancements are implemented, CI will apply the patches against the Kafka generator source, run the patched generator to produce enhanced `*Data` source, compile, and shade the result alongside the rest of `kafka-clients`.

Reviewer 2: Not sure if I have understood this right.  I think the kafka-patches exist to allow us to add features to the Kafka code generator (field aliasing, for example).  Have I groked this right?  

Patching can get tricky and I'd worry about having it  in our foundations.  

I prefer the approach of forking the generator once and owning the forked code going forward.  We'd rely on the serialisation/deserialisation fidelity tests to catch regressions between ourselves and Kafka.

Author: > I think the kafka-patches exist to allow us to add features to the Kafka code generator...

Yes, they would also allow us to bypass the upstream changes we don't like (e.g. we add overloads for functions where the signature has changed).

> I prefer the approach of forking the generator...

I worry that if we keep excluding upstream changes that don't affect the (de)serialized output (e.g. fixes/refactorings), eventually there may be an upstream change that both does affect the output and depends on a series of previous non-output-affecting changes, which would make the update significantly harder. I believe the proposed shading + patches approach helps mitigate this (and drift) in the best way.

Reviewer 2: Resolving - we are no longer considering patches.

Published artifact versions use the Kafka version as a base with a qualifier suffix (e.g. `4.3.0-kroxy-0`, `4.3.0-kroxy-1` for Kroxylicious-only fixes, `4.4.0-kroxy-0` when bumping to a new Kafka release). This makes the Kafka baseline immediately visible in any dependency declaration.

> The `kroxylicious-krpc-plugin` currently depends on `kafka-clients` solely for `ApiKeys`. After this change it depends on the new artifact instead. Its role in generating filter interfaces, invokers, and decoders is otherwise unchanged.

Reviewer 2: Presumably, we could codegen a ApiKeys from the RPC definitions themselves.

Author: ```suggestion
The `kroxylicious-krpc-plugin` currently depends on `kafka-clients` solely for `ApiKeys`. After this change it depends on the new artifact instead (or we could codegen an `ApiKeys` from the RPC definitions themselves). Its role in generating filter interfaces, invokers, and decoders is otherwise unchanged.
```

### Shading configuration

> The shading configuration in `kroxylicious/kafka-patches` uses a single broad relocation rule:

Reviewer 2: So would we release kafka-patches in lock step with kafka releases.  If a kafka release is made which doesn't change wire level API (e.g. 4.3.1) would we still release it?

Author: Presumably, the proposed GitHub workflow creates a PR bumping the version, and all the tests pass straight away, so we just merge it and have another workflow automatically create the patch release. If there are conflicts, I suppose we could optionally defer until a minor release if we're busy?

Reviewer 4: comment is now outdated by latest content

Reviewer 4: we've moved away from shading, closing thread

- **Stability rule:** `org.apache.kafka.common.record.internal.MemoryRecords` → `io.kroxylicious.kafka.common.record.MemoryRecords`

Filter developers who import `io.kroxylicious.kafka.common.record.MemoryRecords` see no change at all. This is the key advantage over a direct `kafka-clients` dependency: Kafka can freely reorganise its internal package structure and the Kroxylicious namespace remains stable. These stability rules accumulate in `kroxylicious/kafka-patches` over time and serve as a permanent, inspectable record of every upstream reorganisation that would otherwise have been a breaking change.

### Verification strategy

The verification harness is a prerequisite, not a follow-up. Correctness must be provable at the point the artifact is first published and on every CI run thereafter.

#### Wire-level round-trip tests

For every spec and every supported protocol version: serialise a message with our generated classes, deserialise with Kafka's, and assert byte-level equality - then reverse the direction. This is the ground truth. If bytes we produce are consumed correctly by Kafka's classes and vice versa, we are wire-compatible.

> Exploratory work on the [`piotrpdev/kroxylicious#refactor/generate-data-class`](https://github.com/piotrpdev/kroxylicious/tree/refactor/generate-data-class) branch built a prototype of this infrastructure covering all 198 specs: parameterised fidelity tests at every valid version, randomised field-value tests, jqwik property-based tests for full-depth message instances, and a compile-time spec count guard. That prototype was designed around the FreeMarker template approach (see ["Rejected Alternatives"](#rejected-alternatives)), where generated classes were compiled at test time via `ToolProvider.getSystemJavaCompiler()` and loaded via `URLClassLoader`. With the shading approach the classes are already on the classpath at build time, so only the class-loading mechanism changes - the verification logic itself carries over unchanged.

Reviewer 2: Where a field is nullable, we'd need to test the null case too.

Author: ```suggestion
Exploratory work on the [`piotrpdev/kroxylicious#refactor/generate-data-class`](https://github.com/piotrpdev/kroxylicious/tree/refactor/generate-data-class) branch built a prototype of this infrastructure covering all 198 specs: parameterised fidelity tests at every valid version, randomised field-value tests (including explicit coverage of `null` for nullable fields), jqwik property-based tests for full-depth message instances, and a compile-time spec count guard. That prototype was designed around the FreeMarker template approach (see ["Rejected Alternatives"](#rejected-alternatives)), where generated classes were compiled at test time via `ToolProvider.getSystemJavaCompiler()` and loaded via `URLClassLoader`. With the shading approach the classes are already on the classpath at build time, so only the class-loading mechanism changes - the verification logic itself carries over unchanged.
```

| New field with `ignorable: true`  | Low        | Safe at all versions                        |

**Proactive early warning:** A scheduled GitHub Actions workflow (e.g. weekly) that detects new Kafka releases and opens a PR in `kroxylicious/kafka-patches` bumping the `kafka-clients` version automatically:

1. Queries Maven Central's search API for the latest release of `org.apache.kafka:kafka-clients`.
2. Compares the result against the current `kafka-clients` version in `kroxylicious/kafka-patches`.
3. If a newer version exists and no open PR already targets it, opens a PR bumping the version.
4. CI runs on the PR - rebuilding the shaded artifact and running the fidelity tests against the new `kafka-clients`.

If CI passes, the PR can be reviewed and merged with minimal effort. If CI fails, the PR surfaces exactly what broke - a fidelity test failure identifies a wire incompatibility; an `EXPECTED_SPEC_COUNT` failure flags a spec change requiring review. Either way, the problem is visible and actionable immediately, before anyone has to manually discover the new release.

#### Drift in shaded classes

> With shading, upstream changes to hand-coded classes like `MemoryRecords` and `RecordBatch` are absorbed automatically when `kafka-clients` is bumped. Drift is detected through two complementary mechanisms:

Reviewer 2: If Kafka changed its MemoryRecord API, we wouldn't necessarily want to ripple the change into our public API surface.

Author: I imagine most changes would be enhancements and fixes, no? I thought we would want to absorb most of the stuff, and we could add patches for reverting what we don't want.

Reviewer 2: In the case of bug fixes, probably yes. 
In the case of API changes, as it would be our API surface, probably not.

Reviewer 4: closing, shading is now a rejected alternative.

### Pure patch-based approach without shading (previous iteration)

Cloning `apache/kafka` at a pinned tag and applying source-level patches to relocate classes before building and publishing. This was the previous iteration of this proposal. It was superseded because changing a Java file's `package` statement requires the file to live in a directory matching its new package for the compiler - meaning every package patch implicitly requires a file move too. This makes the approach a transformation pipeline rather than simple patches, with complexity that shading handles automatically at the bytecode level.

### Fork the `apache/kafka` repository into the Kroxylicious GitHub org

> Maintaining a `kroxylicious/kafka` fork with Kroxylicious-specific commits on top of upstream. This gives a standard Git workflow (edit files, commit, push) and IDE support, and Git's merge/rebase machinery handles context drift more gracefully than patches do. However, it adds a repository to the org that must be kept in sync, the fork's diff from upstream is implicit rather than inspectable, and there is a risk of the fork accumulating unintentional drift. Storing generator patches in `kroxylicious/kafka-patches` achieves the same generator-modification capability with less overhead.

Reviewer 2: When we were talking about forking, I understood that to mean 'just' the generator and the support classes (MemoryRecords, ApiMessage etc).  I don't think any of us imagined forking the whole kafka repo.

Author: By "forking" I thought we meant GitHub forking, apologies. Wouldn't forking the whole repo on GitHub allow us to track changes in the files we want more easily?

Reviewer 2: In this context, I took forking to mean "copy the code" into our repo, rather than depending on it from theirs.

Reviewer 4: thread is now outdated and solution has evolved, closing

interface ApiRequestMessage<RS extends ApiResponseMessage> extends ApiMessage {}
interface ApiResponseMessage extends ApiMessage {}

class ProduceRequestData implements ApiRequestMessage<ProduceResponseData> {}
```

This makes `FilterContext.sendRequest()` type-safe - `context.sendRequest(header, produceRequest)` returns `CompletionStage<ProduceResponseData>` with no cast required. The pairing information already exists in `ApiKeys`; emitting the `implements` clause is a small generator change. It is not required for correctness and can be deferred if it complicates the initial delivery, but the decision to include or exclude it must be made upfront.

#### Post-parity improvements

These can be delivered after the initial release:

- **Javadoc from IDL `about` strings:** The JSON specs already include documentation strings on most fields. The generator can emit these as Javadoc on getters/setters.
- **`aliases` field in the IDL:** An optional field on spec field definitions (e.g. `"aliases": ["oldFieldName"]`) allowing Kroxylicious to maintain stable getter/setter names even if Kafka renames a field in a future IDL version. Requires schema model changes and generator work.

> ### Migration for filter developers

Reviewer 2: Non-blocking - @SamBarker would this be a good candidate for OpenRewrite?  I'm not suggesting we need this in the proposal.
